src/Http/Response.php

```diff
@@ -66,4 +118,70 @@ public static function createSwitchProtocolResponse()
 
         return $response;
     }
+
+    /**
+     * @param string $data
+     * @return Response
+     * @throws HttpException
+     */
+    public static function create(string &$data) : Response
+    {
+        if (!\preg_match('/\\r\\n\\r\\n/', $data)) {
+            throw new IncompleteHttpMessageException();
+        }
+
+        // Split response headers from potential content
+        $exploded = explode("\r\n\r\n", $data);
+        $responseString = '';
+
+        if (count($exploded) > 1) {
+            // Removes the request keep content in data reference
+            $responseString = $exploded[0];
+            unset($exploded[0]);
+            $data = implode("\r\n\r\n", $exploded);
+        }
+
+        $response = new Response();
+
+        $lines = \explode("\r\n", $responseString);
+        Response::initResponse($lines[0], $response);
+
+        unset($lines[0]);
+        Response::initHeaders($lines, $response);
+
+        if (strtolower($response->getHeader('Upgrade')) !== 'websocket') {
+            throw new HttpException('Missing or wrong upgrade header.');
+        }
+        if (strtolower($response->getHeader('Connection')) !== 'upgrade') {
+            throw new HttpException('Missing upgrade header.');
```

Comment:
The exception message is a bit misleading. It's actually the **`Connection`** header that is whether missing or incorrect.
Could be even more explicit to indicate that this header should be set to `upgrade` 🤔